elide-datastore/elide-datastore-hibernate3/src/main/java/com/yahoo/elide/datastores/hibernate3/HibernateTransaction.java

```diff
@@ -188,15 +201,134 @@ public void commit() {
             final Pagination paginationData = pagination.get();
             sessionCriteria.setFirstResult(paginationData.getOffset());
             sessionCriteria.setMaxResults(paginationData.getLimit());
+        } else {
+            Integer queryLimit = getQueryLimit();
+            if (queryLimit != null) {
+                sessionCriteria.setMaxResults(queryLimit);
+            }
+        }
+
+        if (isJoinQuery()) {
+            joinCriteria(sessionCriteria, loadClass);
         }
 
         @SuppressWarnings("unchecked")
-        Iterable<T> list = new ScrollableIterator(sessionCriteria.scroll(ScrollMode.FORWARD_ONLY));
+        Iterable<T> list;
+        if (isJoinQuery()) {
+            list = sessionCriteria.list();
+        } else {
+            list = new ScrollableIterator(sessionCriteria.scroll(ScrollMode.FORWARD_ONLY));
+        }
 
         return list;
     }
 
     /**
+     * Should this transaction use JOINs. By default will be enabled if Sparse fields are available.
+     *
+     * @return true to use join logic
+     */
+    public boolean isJoinQuery() {
+        if (requestScope != null) {
+            if (!requestScope.getSparseFields().isEmpty()) {
+                return true;
+            }
+
+            if (requestScope.getQueryParams().isPresent()
+                    && requestScope.getQueryParams().get().get("include") != null) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private <T> void joinCriteria(Criteria criteria, final Class<T> loadClass) {
+        EntityDictionary dictionary = requestScope.getDictionary();
+        String type = dictionary.getBinding(loadClass);
+        Set<String> fields = Objects.firstNonNull(
+                requestScope.getSparseFields().get(type), Collections.<String>emptySet());
+        List<String> relationships = dictionary.getRelationships(loadClass);
+        for (String field : fields) {
+            try {
+                checkFieldReadPermission(loadClass, field);
```

Comment:
Why are we doing security checks inside the data store?  These should really be isolated to persistent resource.